sockeye/layers.py

```diff
@@ -46,11 +48,1542 @@ def activation(data: mx.sym.Symbol, act_type: str) -> mx.sym.Symbol:
+
+    def att_names(self):
+        att_names = []
+        for layer in self.layers:
+            att_names.extend(layer.att_names())
+        return att_names
+
+    def get_num_hidden(self) -> int:
+        return self.layers[-1].get_num_hidden()
+
+
+class DecoderLayerChain(NestedDecoderLayer):
+
+    def __init__(self, layers: List[DecoderLayer]):
+        self._layers = layers
+
+    @property
+    def layers(self) -> List[DecoderLayer]:
+        return self._layers
+
+    def decode_sequence(self, source_encoded: mx.sym.Symbol, source_encoded_lengths: mx.sym.Symbol,
+                        source_encoded_max_length: int, target_encoded: mx.sym.Symbol,
+                        target_encoded_lengths: mx.sym.Symbol, target_encoded_max_length: int,
+                        target_autoregressive_bias: mx.sym.Symbol) -> mx.sym.Symbol:
+        for layer in self.layers:
+            target_encoded = layer.decode_sequence(source_encoded,
+                                                   source_encoded_lengths,
+                                                   source_encoded_max_length,
+                                                   target_encoded,
+                                                   target_encoded_lengths,
+                                                   target_encoded_max_length,
+                                                   target_autoregressive_bias)
+
+        return target_encoded
+
+    def decode_step(self, step: int, source_encoded: mx.sym.Symbol, source_encoded_lengths: mx.sym.Symbol,
+                    source_encoded_max_length: int, target: mx.sym.Symbol, layer_states_flat: List[mx.sym.Symbol], att_dict) -> Tuple[mx.sym.Symbol, List[mx.sym.Symbol]]:
+        new_layer_states_flat = []
+        for layer, layer_states in self.layers_with_states_iter(step, layer_states_flat):
+            target, new_layer_states = layer.decode_step(step, source_encoded, source_encoded_lengths,
+                                                         source_encoded_max_length, target, layer_states, att_dict)
+            new_layer_states_flat.extend(new_layer_states)
+
+        return target, new_layer_states_flat
+
+    def get_num_hidden(self) -> int:
+        return self._layers[-1].get_num_hidden()
+
+
+class StatelessBlock(ABC):
+    """
+    A block which does not require to keep any state during at inference time so that we are able to call it one
+    timestep at a time. This basically means that the block can be run as
+
+        concat(block(data[:,t,:]) for t in range(seq_len), dim=1)
+
+    Namely, at time t we only need the data point at time t from the previous layer.
+    """
+
+    @abstractmethod
+    def __call__(self,
+                 data: mx.sym.Symbol,
+                 lengths: Optional[mx.sym.Symbol] = None,
+                 max_length: Optional[int] = None) -> mx.sym.Symbol:
+        """
+        Compute the block for the entire sequence.
```

Comment:
" for the entire sequence data, shaped (batch_size, sequence length, num_hidden)."

Reply:
The shape is already mentioned in the `:param` block below. I'd rather not make this redundant.